driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -149,6 +156,7 @@ public InternalConnection get(final long timeoutValue, final TimeUnit timeUnit)
         connectionPoolListener.connectionCheckOutStarted(new ConnectionCheckOutStartedEvent(serverId));
         Timeout timeout = Timeout.startNow(timeoutValue, timeUnit);
         try {
+            stateAndGeneration.throwIfClosedOrPaused();
```

Comment:
There are two parts to address here:

1. The closed state has previously checked here anyway, but only in the `ConcurrentPool`. So effectively nothing has changed with regard to this state.
2. Yes, in situations when a server is selected for checkout and then paused (as a result of a problem when either monitoring or using it), an application may observe a `MongoConnectionPoolClearedException`. This does increase the probability of an application observing failures (strictly speaking, this is true even if we retry these exceptions on our own, though in a much less extent).

Reply:
Let's discuss this one.

Reply:
@jyemin I did conducted the experiment we discussed, and indeed, the usage scenario in the experiment succeeds on server version 4.2.14 and fails on 4.0.25.

### A short description of the experiment

- Check whether an operation that is queued while checking out a connection from `DefaultConnectionPool` fails or succeeds as a result of failover.
- Servers with versions before 4.2 terminate all connections when stepping down, the driver accordingly invalidates the pool (and pauses it with the changes made in this PR). Servers with versions starting at 4.2 do not terminate connections when stepping down and the driver does not invalidate the pool.
- The expectation is that the aforementioned operation fails on < 4.2 with `MongoConnectionPoolClearedException` and succeeds on >= 4.2.

### Reproducer

1. `FailoverExperiment.java`:

```java
package com.mongodb.client;

import com.mongodb.MongoClientSettings;
import com.mongodb.ServerAddress;
import com.mongodb.connection.ClusterConnectionMode;
import com.mongodb.connection.ClusterType;
import org.bson.BsonBoolean;
import org.bson.BsonDocument;
import org.bson.BsonInt64;
import org.bson.Document;

import java.time.LocalTime;
import java.util.Collections;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

final class FailoverExperiment {
    public static void main(final String[] args) throws InterruptedException, ExecutionException, TimeoutException {
        final String appName = FailoverExperiment.class.getSimpleName();
        MongoClientSettings.Builder clientSettingsBuilder = MongoClientSettings.builder()
                .applicationName(appName)
                .applyToClusterSettings(builder -> builder
                        .mode(ClusterConnectionMode.MULTIPLE)
                        .requiredClusterType(ClusterType.REPLICA_SET)
                        .hosts(Collections.singletonList(new ServerAddress(ServerAddress.defaultHost(), ServerAddress.defaultPort()))))
                .applyToConnectionPoolSettings(builder -> builder
                        .maxSize(1)
                        .maxWaitTime(1, TimeUnit.MINUTES));
        try (MongoClient client = MongoClients.create(clientSettingsBuilder.applicationName(appName).build());
             MongoClient adminClient = MongoClients.create(clientSettingsBuilder.applicationName(appName + " admin").build())) {
            MongoDatabase db = client.getDatabase(appName + "Db");
            MongoCollection<Document> collection = db.getCollection(appName + "Collection");
            Future<?> find1 = findAsync(collection, "find1");
            Future<?> find2 = findAsync(collection, "find2");
            // wait until one of `find`s checks out a connection and the other one begins waiting for a connection
            Thread.sleep(2000);
            replSetStepDown(adminClient.getDatabase("admin"), 10);
            find1.get(1, TimeUnit.MINUTES);
            find2.get(1, TimeUnit.MINUTES);
        }
    }

    private static void replSetStepDown(final MongoDatabase adminDb, final long replSetStepDownSeconds) {
        try {
            log("replSetStepDown result " + adminDb.runCommand(
                    new BsonDocument("replSetStepDown", new BsonInt64(replSetStepDownSeconds))
                            .append("force", BsonBoolean.TRUE)));
        } catch (final RuntimeException e) {
            log("replSetStepDown result " + e);

        }
    }

    private static Future<?> findAsync(final MongoCollection<?> collection, final String id) {
        return ForkJoinPool.commonPool().submit(() -> {
            try {
                log(id + " begin");
                log(id + " result " + collection.find().first());
            } catch (Exception e) {
                log(id + " failed");
                e.printStackTrace(System.out);
            }
        });
    }

    private static void log(final Object msg) {
        System.out.printf("%s [%s] %s - %s%n",
                LocalTime.now(), Thread.currentThread().getName(), FailoverExperiment.class.getSimpleName(), msg);
    }
}
```

2. Since [`blockConnection` mode for `failCommand` is supported only starting at 4.2.9](https://jira.mongodb.org/browse/SERVER-41070), the following modification to the driver code was made to emulate a long operation: `sleepFirstNTimes(7_000);` is inserted immediately after the [line where a `find` operation checks out a connection](https://github.com/mongodb/mongo-java-driver/blob/3196681720bb4e5ec43ed10cabf270831d34b1c1/driver-core/src/main/com/mongodb/internal/operation/FindOperation.java#L655). `sleepFirstNTimes` is defined as follows:
```java
private static final AtomicInteger SLEEP_FIRST_N_TIMES = new AtomicInteger(1);

private static void sleepFirstNTimes(final long millis) {
    if (SLEEP_FIRST_N_TIMES.getAndDecrement() > 0) {
        try {
            Thread.sleep(millis);
        } catch (final InterruptedException e) {
            throw new MongoInterruptedException(null, e);
        }
    }
}
```

### Results

#### 4.0.25
In this execution `find2` was the first operation that checked out a connection, so `find1` had to be queued while waiting for a connection, and it failed with `MongoConnectionPoolClearedException`.

```
16:15:57.742314 [ForkJoinPool.commonPool-worker-19] FailoverExperiment - find1 begin
16:15:57.742361 [ForkJoinPool.commonPool-worker-5] FailoverExperiment - find2 begin
...
16:15:59.771489 [main] FailoverExperiment - replSetStepDown result com.mongodb.MongoSocketReadException: Prematurely reached end of stream
16:16:04.813308 [ForkJoinPool.commonPool-worker-19] FailoverExperiment - find1 failed
com.mongodb.MongoConnectionPoolClearedException: Connection pool for ServerId{clusterId=ClusterId{value='60d10f9849e16346a55b9ca0', description='null'}, address=localhost:27019} is paused because another operation failed
...
Caused by: com.mongodb.MongoSocketReadException: Prematurely reached end of stream
...
16:16:09.383229 [ForkJoinPool.commonPool-worker-5] FailoverExperiment - find2 result null
```

#### 4.2.14
In this execution `find2` again was the first operation that checked out a connection, so `find1` had to be queued while waiting for a connection, but it succeeded once `find2` succeeded and released its connection.

```
16:22:56.245403 [ForkJoinPool.commonPool-worker-5] FailoverExperiment - find2 begin
16:22:56.245542 [ForkJoinPool.commonPool-worker-19] FailoverExperiment - find1 begin
...
16:22:58.270983 [main] FailoverExperiment - replSetStepDown result Document{{ok=1.0, $clusterTime=Document{{clusterTime=Timestamp{value=6976376251464286209, seconds=1624314173, inc=1}, signature=Document{{hash=org.bson.types.Binary@c98f581, keyId=0}}}}, operationTime=Timestamp{value=6976376251464286209, seconds=1624314173, inc=1}}}
...
16:23:06.858401 [ForkJoinPool.commonPool-worker-5] FailoverExperiment - find2 result null
16:23:06.860280 [ForkJoinPool.commonPool-worker-19] FailoverExperiment - find1 result null
```